src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/authorization/ServiceProviderAccessScopeMapper.kt

```diff
@@ -34,63 +40,81 @@ class ServiceProviderAccessScopeMapper(
 
-    val serviceProvider = getServiceProvider(serviceProviderGroups, configErrors)
+    val providers = getProviders(serviceProviderGroups, scope.errors)
+    scope.providers.addAll(providers)
+  }
 
-    val contractGroups = groups
+  private fun resolveContracts(scope: WorkingScope) {
+    val contractGroups = scope.authGroups
       .filter { it.startsWith(contractGroupPrefix) }
       .map { it.removePrefix(contractGroupPrefix) }
 
-    val contracts = getContracts(contractGroups, configErrors)
+    val contracts = getContracts(contractGroups, scope.errors)
+    scope.contracts.addAll(contracts)
+  }
 
-    // FIXME we also need to remove contracts which do not belong to the user's provider
+  private fun removeInaccessibleContracts(scope: WorkingScope) {
+    val orphanContracts = scope.contracts.filterNot {
+      scope.providers.contains(it.primeProvider) || it.subcontractorProviders.intersect(scope.providers).isNotEmpty()
+    }
+    orphanContracts.forEach {
+      scope.errors.add("contract '${it.contractReference}' is not accessible to providers ${scope.providers.map { p -> p.id }}")
+    }
+    scope.contracts.removeAll(orphanContracts)
+  }
 
-    if (configErrors.isNotEmpty()) {
-      throw AccessError(user, errorMessage, configErrors)
+  private fun blockUsersWithoutContracts(scope: WorkingScope) {
+    if (scope.contracts.isEmpty()) {
+      scope.errors.add("no valid contract groups associated with user")
     }
+  }
 
-    // this not null assertion on serviceProvider is ugly, but it's the only way i could think
-    // of to allow all the errors to be processed in a sensible way above.
-    return ServiceProviderAccessScope(serviceProvider = serviceProvider!!, contracts = contracts)
+  private fun blockUsersWithoutProviders(scope: WorkingScope) {
+    if (scope.providers.isEmpty()) {
+      scope.errors.add("no valid service provider groups associated with user")
+    }
   }
 
-  private fun getServiceProvider(serviceProviderGroups: List<String>, configErrors: MutableList<String>): ServiceProvider? {
-    return when {
-      serviceProviderGroups.isEmpty() -> {
-        configErrors.add("no service provider groups associated with user")
-        null
-      }
-      serviceProviderGroups.size > 1 -> {
-        configErrors.add("more than one service provider group associated with user")
-        null
-      }
-      else -> {
-        val providerGroupCode = serviceProviderGroups[0]
-        val provider = serviceProviderRepository.findByIdOrNull(providerGroupCode)
-        if (provider == null) {
-          configErrors.add("service provider id '$providerGroupCode' does not exist in the interventions database")
-        }
-        provider
-      }
+  private fun getProviders(providerGroups: List<String>, configErrors: MutableList<String>): List<ServiceProvider> {
+    val providers = serviceProviderRepository.findAllById(providerGroups)
+    val removedProviders = providerGroups.subtract(providers.map { it.id })
```

Comment:
i think here and in `getContracts` the term 'removed' is a little misleading. nothing is getting removed, these are just sets of providers and contracts that exist in the user's auth config, but not in the interventions database. so these are more like 'unknownProviders' or 'unresolvedProviders' or something. then the error message could be something like "provider group {} was not added to access scope because it does not exist in the refer and monitor database"

Reply:
done in 94cb6204470d325c967d5340b6914d0c1a8a1c0a -- I opted for "unidentified"